clang/test/SemaSYCL/non-kernel-functor.cpp

```diff
@@ -0,0 +1,21 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown  -verify %s
+
+// Test to verify that non-kernel functors are not processed as SYCL kernel
+// functors
+
+// expected-no-diagnostics
```

Comment:
Just curious, what is the error was emitted for this code?

Reply:
I don't believe it was an error, but it hit the assertion in SemaSYCL.cpp:830

Reply:
The AST here was - 

```
FunctionDecl 0x1042b0e8 <<built-in>:796:23, test.cpp:18:1> line:15:6 foo 'void ()'
|-CompoundStmt 0x1042b9a0 <col:12, line:18:1>
| |-DeclStmt 0x1042b810 <line:16:3, col:16>
| | `-VarDecl 0x1042b238 <col:3, col:10> col:10 used m_uold 'Second' callinit
| |   `-CXXConstructExpr 0x1042b7e8 <col:10> 'Second' 'void () noexcept'
| `-ExprWithCleanups 0x1042b988 <line:17:3, col:12> 'void'
|   `-CXXOperatorCallExpr 0x1042b958 <col:3, col:12> 'void' '()'
|     |-ImplicitCastExpr 0x1042b928 <col:11, col:12> 'void (*)()' <FunctionToPointerDecay>
|     | `-DeclRefExpr 0x1042b8e0 <col:11, col:12> 'void ()' lvalue CXXMethod 0x103d9f08 'operator()' 'void ()'
|     `-MaterializeTemporaryExpr 0x1042b940 <col:3, col:10> 'First' lvalue
|       `-CXXOperatorCallExpr 0x1042b8b0 <col:3, col:10> 'First' '()'
|         |-ImplicitCastExpr 0x1042b898 <col:9, col:10> 'First (*)()' <FunctionToPointerDecay>
|         | `-DeclRefExpr 0x1042b848 <col:9, col:10> 'First ()' lvalue CXXMethod 0x103da238 'operator()' 'First ()'
|         `-DeclRefExpr 0x1042b828 <col:3> 'Second' lvalue Var 0x1042b238 'm_uold' 'Second'
`-SYCLDeviceAttr 0x1042b188 <<built-in>:796:38>
```

As you can see there are 2 operator calls when `CallStack `is 1 (i.e. when `foo `is in `CallStack`). This caused the assert to get hit.